File: zstacktestagent/zstacktestagent/plugins/vm.py
'''

@author: Frank
'''

import zstacktestagent.testagent as testagent
import zstacklib.utils.shell as shell
import zstacklib.utils.jsonobject as jsonobject
import zstacklib.utils.http as http
import zstacklib.utils.linux as linux
import zstacklib.utils.ssh as ssh
import zstacklib.utils.log as log
import threading
import time

logger = log.get_logger(__name__)

class VmStatusCmd(testagent.AgentCommand):
    def __init__(self):
        super(VmStatusCmd, self).__init__()
        self.vm_uuids = None

class VmStatusRsp(testagent.AgentResponse):
    def __init__(self):
        super(VmStatusRsp, self).__init__()
        self.vm_status = {}

class DeleteVmCmd(testagent.AgentCommand):
    def __init__(self):
        super(DeleteVmCmd, self).__init__()
        self.vm_uuids = None

class SshInVmCmd(testagent.AgentCommand):
    def __init__(self):
        super(SshInVmCmd, self).__init__()
        self.ip = None
        self.username = None
        self.password = None
        self.port = 22
        self.timeout = 180 # seconds
        self.command = None

class SshInVmRsp(testagent.AgentResponse):
    def __init__(self):
        super(SshInVmRsp, self).__init__()
        self.result = None

class ScpInVmCmd(testagent.AgentCommand):
    def __init__(self):
        super(ScpInVmCmd, self).__init__()
        self.ip = None
        self.username = None
        self.password = None
        self.port = 22
        self.timeout = 180 # seconds
        self.src_file = None
        self.dst_file = None

class ScpInVmRsp(testagent.AgentResponse):
    def __init__(self):
        super(ScpInVmRsp, self).__init__()
        self.result = None

         
IS_VM_STOPPED_PATH = '/vm/isvmstopped'
IS_VM_DESTROYED_PATH = '/vm/isvmdestroyed'
IS_VM_RUNNING_PATH = '/vm/isvmrunning'
DELETE_VM_PATH = '/vm/deletevm'
SSH_GUEST_VM_PATH = '/vm/sshguestvm'
SCP_GUEST_VM_PATH = '/vm/scpguestvm'
VM_STATUS = '/vm/vmstatus'
LIST_ALL_VM = '/vm/listallvm'
VM_BLK_STATUS = '/vm/vmblkstatus'
ECHO_PATH = '/host/echo'
        
class VmAgent(testagent.TestAgent):
    VM_STATUS_RUNNING = 'running'
    VM_STATUS_STOPPED = 'shut off'
    VM_STATUS_DESTROYED = None
    VM_EXCEPTION_STATUS = 'EXCEPTION_STATUS'
    
    def start(self):
        testagent.TestAgentServer.http_server.register_sync_uri(IS_VM_RUNNING_PATH, self.is_vm_running)
        testagent.TestAgentServer.http_server.register_sync_uri(IS_VM_DESTROYED_PATH, self.is_vm_stopped)
        testagent.TestAgentServer.http_server.register_sync_uri(IS_VM_STOPPED_PATH, self.is_vm_stopped)
        testagent.TestAgentServer.http_server.register_sync_uri(DELETE_VM_PATH, self.delete_vm)
        testagent.TestAgentServer.http_server.register_sync_uri(SSH_GUEST_VM_PATH, self.ssh_in_guest_vm)
        testagent.TestAgentServer.http_server.register_sync_uri(SCP_GUEST_VM_PATH, self.scp_in_guest_vm)
        testagent.TestAgentServer.http_server.register_sync_uri(VM_STATUS, self.get_vm_status)
        testagent.TestAgentServer.http_server.register_sync_uri(LIST_ALL_VM, self.list_all_vms)
        testagent.TestAgentServer.http_server.register_sync_uri(VM_BLK_STATUS, self.get_vm_blk_status)
        testagent.TestAgentServer.http_server.register_sync_uri(ECHO_PATH, self.echo)
        shell.logcmd = True

    @testagent.replyerror
    def echo(self, req):
        logger.debug('echo ping')
        return ''
    
    def stop(self):
        pass
    
    def _list_all_vms(self):
        output = shell.call('virsh list --all')
        return output.split('\n')
    
    def _is_vm_status(self, uuid, status):
        curr_status = self._get_vm_status(uuid)
        if status:
            if status in curr_status:
                return True
        else:
            if curr_status != self.VM_EXCEPTION_STATUS:
                return True
        logger.debug('[vm uuid: %s] does not have status: %s.' % (uuid, status))
        return False

    def _get_vm_status(self, uuid):
        try:
            output = shell.call('virsh domstate %s' % uuid)
        except Exception as e:
            logger.debug('Exception happened when trying to get [vm uuid: %s] status' % uuid)
            return self.VM_EXCEPTION_STATUS
            
        return output

    def _vm_blk_status(self, uuid):
        output = shell.call('virsh domblklist %s' % uuid).split('\n')
        output = output[2:]
        ret = {}
        for item in output:
            if item != '':
                blk = item.split()
                ret[blk[0]] = blk[1]
        return ret

    def _delete_vm(self, uuid):
        shell.call('virsh undefine --managed-save %s' % uuid)
    
    def _destroy_vm(self, uuid):
        shell.call('virsh destroy %s' % uuid)
    
    def _delete_all_vm(self):
        output = self._list_all_vms()
        output = filter(bool, output)
        for o in output:
            uuid = o.split()[1]
            if self.VM_STATUS_RUNNING in o:
                self._destroy_vm(uuid)
                self._delete_vm(uuid)
            else:
                self._delete_vm(uuid)
    
    @testagent.replyerror
    def is_vm_running(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        rsp = VmStatusRsp()
        for uuid in cmd.vm_uuids:
            rsp.vm_status[uuid] = self._is_vm_status(uuid, self.VM_STATUS_RUNNING)
        return jsonobject.dumps(rsp)
    
    @testagent.replyerror
    def is_vm_stopped(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        rsp = VmStatusRsp()
        for uuid in cmd.vm_uuids:
            rsp.vm_status[uuid] = self._is_vm_status(uuid, self.VM_STATUS_STOPPED)
        return jsonobject.dumps(rsp)
    
    @testagent.replyerror
    def is_vm_destroyed(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        rsp = VmStatusRsp()
        for uuid in cmd.vm_uuids:
            rsp.vm_status[uuid] = self._is_vm_status(uuid, self.VM_STATUS_DESTROYED)
        return jsonobject.dumps(rsp)
    
    @testagent.replyerror
    def get_vm_status(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        rsp = VmStatusRsp()
        for uuid in cmd.vm_uuids:
            rsp.vm_status[uuid] = self._get_vm_status(uuid)
            logger.debug('[vm:%s status:] %s.' % (uuid, rsp.vm_status[uuid]))
        return jsonobject.dumps(rsp)
    
    @testagent.replyerror
    def get_vm_blk_status(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        rsp = VmStatusRsp()

        for uuid in cmd.vm_uuids:
            rsp.vm_status[uuid] = self._vm_blk_status(uuid)
        return jsonobject.dumps(rsp)

    @testagent.replyerror
    def list_all_vms(self, req):
        rsp = VmStatusRsp()
        rsp.vm_status['vms'] = self._list_all_vms()
        return jsonobject.dumps(rsp)

    @testagent.replyerror
    def delete_vm(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        vmuuids = cmd.vm_uuids_
        if not vmuuids:
            self._delete_all_vm()
        else:
            for uuid in vmuuids:
                if self._is_vm_status(uuid, self.VM_STATUS_RUNNING):
                    self._destroy_vm(uuid)
                if (self._is_vm_status(uuid, self.VM_STATUS_STOPPED) or self._is_vm_status(uuid, self.VM_STATUS_DESTROYED)):
                    self._delete_vm(uuid)
        return jsonobject.dumps(testagent.AgentResponse())

    @testagent.replyerror
    def ssh_in_guest_vm(self, req):
        rsp = SshInVmRsp()
        rsp_dict = {'error': None, 'completion': None}
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        
        def login_vm():
            try:
                ret, output, stderr = ssh.execute(cmd.command, cmd.ip, cmd.username, cmd.password, False)
                if ret != 0:
                    rsp.success = False
                    rsp.error = '%s\n%s' % (output, stderr)
                else:
                    rsp.result = output

                rsp_dict['completion'] = True
                return True

            except Exception as e:
                logger.debug('[SSH] unable to ssh in vm[ip:%s], assume its not ready. Exception: %s' % (cmd.ip, str(e)))
                rsp_dict['error'] = True
                rsp_dict['completion'] = True
                
            return False

        thread = threading.Thread(target = login_vm)
        thread.start()
        timeout = time.time() + cmd.timeout 
        while not rsp_dict['completion'] and time.time() < timeout:
            time.sleep(0.5)

        if rsp_dict['completion']:
            if rsp_dict['error']:
                rsp.success = False
                rsp.error = 'ssh command:%s met exception.' % cmd.command
                logger.debug('ssh command:%s met exception.' % cmd.command)
        else:
            logger.debug('[SSH] ssh in vm[%s] doing %s, timeout after %s seconds' % (cmd.ip, cmd.command, cmd.timeout))
            rsp.success = False
            rsp.error = 'ssh execution keeps failure, until timeout: %s' \
                % cmd.timeout

        logger.debug('[SSH] ssh in vm[%s] doing %s done. result is %s' % (cmd.ip, cmd.command, rsp.success))
        return jsonobject.dumps(rsp)

    @testagent.replyerror
    def scp_in_guest_vm(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        rsp = ScpInVmRsp()
        
        try:
            ssh.scp_file(cmd.src_file, cmd.dst_file, cmd.ip, cmd.username, cmd.password, cmd.port)
            rsp.success = True
            rsp.output = '[SCP] Successfully scp %s to [vm:] %s %s' % \
                    (cmd.src_file, cmd.ip, cmd.dst_file)

        except Exception as e:
            logger.debug('[SCP] scp %s to vm[ip:%s] failed: %s.' % \
                    (cmd.src_file, cmd.ip, str(e)))
            rsp.success = False
            rsp.error = str(e)

        return jsonobject.dumps(rsp)

File: zstacktestagent/testagent/src/testagent/plugin/vm.go
package plugin

import (
	"fmt"
	"strings"
	"testagent/server"
	"testagent/utils"
	"time"
)

type VmStatusCheckCmd struct {
	Uuids []string `json:"vm_uuids"`
}

type VmDiskQosCmd struct {
	Uuid string `json:"vm_uuid"`
	Dev  string `json:"dev"`
}

type VmSshGuestVmCmd struct {
	Ip       string `json:"ip"`
	Command  string `json:"command"`
	Username string `json:"username"`
	Password string `json:"password"`
	Timeout  int    `json:"timeout"`
}

type VmSshGuestVmRsp struct {
	Success bool   `json:"success"`
	Error   string `json:"error"`
	Result  string `json:"result"`
}

type VmScpGuestVmCmd struct {
	Ip       string `json:"ip"`
	Username string `json:"username"`
	Password string `json:"password"`
	SrcFile  string `json:"src_file"`
	DstFile  string `json:"dst_file"`
	Timeout  int    `json:"timeout"`
}

type VmScpGuestVmRsp struct {
	Success bool   `json:"success"`
	Error   string `json:"error"`
}

type VmDiskQosRsp struct {
	VmDeviceIo string `json:"vm_device_io"`
}

type VmStatusCheckRsp struct {
	Vm_status map[string]bool `json:"vm_status"`
}

type VmStatusRsp struct {
	Vm_status map[string]string `json:"vm_status"`
}

type VmListAllVmRsp struct {
	Vm_status map[string][]string `json:"vm_status"`
}

type VmBlkStatusRsp struct {
	Vm_status map[string]map[string]string `json:"vm_status"`
}

const (
	IS_VM_STOPPED_PATH   = "/vm/isvmstopped"
	IS_VM_PAUSED_PATH    = "/vm/isvmpaused"
	IS_VM_DESTROYED_PATH = "/vm/isvmdestroyed"
	IS_VM_RUNNING_PATH   = "/vm/isvmrunning"
	SSH_GUEST_VM_PATH    = "/vm/sshguestvm"
	SCP_GUEST_VM_PATH    = "/vm/scpguestvm"
	VM_STATUS            = "/vm/vmstatus"
	LIST_ALL_VM          = "/vm/listallvm"
	VM_BLK_STATUS        = "/vm/vmblkstatus"
	VM_ECHO_PATH         = "/host/echo"
	VM_DEVICE_QOS        = "/vm/deviceqos"
)

const (
	VM_STATUS_RUNNING   = "running"
	VM_STATUS_STOPPED   = "shut off"
	VM_STATUS_PAUSED    = "paused"
	VM_STATUS_DESTROYED = ""
	VM_EXCEPTION_STATUS = "EXCEPTION_STATUS"
)

func isVmStatus(vm_uuid string, status string) bool {
	currentStatus := vmStatus(vm_uuid)
	fmt.Printf("%s\n", currentStatus)
	if currentStatus == status {
		return true
	} else {
		return false
	}
}

func vmStatus(vm_uuid string) string {
	bash := utils.Bash{
		Command: fmt.Sprintf("virsh domstate %s", vm_uuid),
		NoLog:   true,
	}

	retCode, o, _, err := bash.RunWithReturn()
	if err != nil || retCode != 0 {
		return VM_EXCEPTION_STATUS
	} else {
		return strings.TrimSpace(o)
	}
}

func vmSerialLog(vm_uuid string, status string) string {
	if status != VM_STATUS_RUNNING {
		_bash := utils.Bash{
			Command: fmt.Sprintf("rm -rf /tmp/%s_serial.log", vm_uuid),
			NoLog:   true,
		}

		_bash.Run()
		return VM_EXCEPTION_STATUS
	}

	bash_pts := utils.Bash{
		Command: fmt.Sprintf("grep pts /var/log/libvirt/qemu/%s.log | awk '{print $9}'", vm_uuid),
		NoLog:   true,
	}

	_retCode, _o, _, _err := bash_pts.RunWithReturn()
	if _err != nil || _retCode != 0 {
		return VM_EXCEPTION_STATUS
	} else {

		bash_serial_log := utils.Bash{
			Command: fmt.Sprintf("sh /tmp/serial_log_gen.sh %s_serial.log %s", vm_uuid, _o),
			NoLog:   true,
		}

		bash_serial_log.Run()
		return vm_uuid + "_serial.log"
	}
}

func vmIsVmStoppedHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmStatusCheckCmd{}
	ctx.GetCommand(cmd)
	vm_status := make(map[string]bool)
	for _, uuid := range cmd.Uuids {
		vm_status[uuid] = isVmStatus(uuid, VM_STATUS_STOPPED)
	}

	return VmStatusCheckRsp{Vm_status: vm_status}
}

func vmIsVmSuspendedHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmStatusCheckCmd{}
	ctx.GetCommand(cmd)
	vm_status := make(map[string]bool)
	for _, uuid := range cmd.Uuids {
		vm_status[uuid] = isVmStatus(uuid, VM_STATUS_PAUSED)
	}

	return VmStatusCheckRsp{Vm_status: vm_status}
}

func vmIsVmDestroyedHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmStatusCheckCmd{}
	ctx.GetCommand(cmd)
	vm_status := make(map[string]bool)
	for _, uuid := range cmd.Uuids {
		vm_status[uuid] = isVmStatus(uuid, VM_STATUS_DESTROYED)
	}

	return VmStatusCheckRsp{Vm_status: vm_status}
}

func vmIsVmRunningHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmStatusCheckCmd{}
	ctx.GetCommand(cmd)
	vm_status := make(map[string]bool)
	for _, uuid := range cmd.Uuids {
		vm_status[uuid] = isVmStatus(uuid, VM_STATUS_RUNNING)
	}

	return VmStatusCheckRsp{Vm_status: vm_status}
}

func vmSshGuestVmHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmSshGuestVmCmd{}
	ctx.GetCommand(cmd)
	var rsp VmSshGuestVmRsp
	timeout := make(chan VmSshGuestVmRsp, 1)
	go func() {
		output, err := utils.ExecuteCommand(cmd.Ip, "22", cmd.Username, cmd.Password, cmd.Command)
		if err != nil {
			rsp.Success = false
			rsp.Error = fmt.Sprintf("[SSH] unable to ssh in vm[ip:%s], assume its not ready. Exception: %s ", cmd.Ip, err)
			rsp.Result = ""
		} else {
			rsp.Success = true
			rsp.Error = ""
			rsp.Result = output
		}
		timeout <- rsp
	}()

	select {
	case rsp = <-timeout:
	case <-time.After(time.Second * time.Duration(cmd.Timeout)):
		rsp.Success = false
		rsp.Error = fmt.Sprintf("ssh execution keeps failure, until timeout: %d", cmd.Timeout)
		rsp.Result = ""
	}

	return rsp
}

func vmScpGuestVmHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmScpGuestVmCmd{}
	ctx.GetCommand(cmd)
	var rsp VmScpGuestVmRsp
	err := utils.ScpFile(cmd.Ip, "22", cmd.Username, cmd.Password, cmd.SrcFile, cmd.DstFile)
	if err != nil {
		rsp.Success = false
		rsp.Error = fmt.Sprintf("[SSH] unable to ssh to vm[ip:%s]. Exception: %s ", cmd.Ip, err)
	} else {
		rsp.Success = true
		rsp.Error = ""
	}

	return rsp
}

func vmStatusHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmStatusCheckCmd{}
	ctx.GetCommand(cmd)
	vm_status := make(map[string]string)
	for _, uuid := range cmd.Uuids {
		vm_status[uuid] = vmStatus(uuid)
		vm_status[uuid+"_log"] = vmSerialLog(uuid, vmStatus(uuid))
	}

	return VmStatusRsp{Vm_status: vm_status}
}

func listAllVm() []string {
	bash := utils.Bash{
		Command: "virsh list --all",
		NoLog:   true,
	}

	_, o, _, _ := bash.RunWithReturn()
	bash.PanicIfError()
	return strings.Split(o, "\n")
}

func vmListAllVmHandler(ctx *server.CommandContext) interface{} {
	vm_status := make(map[string][]string)
	vm_status["vms"] = listAllVm()
	return VmListAllVmRsp{Vm_status: vm_status}
}

func vmBlkStatus(uuid string) map[string]string {
	bash := utils.Bash{
		Command: fmt.Sprintf("virsh domblklist %s", uuid),
		NoLog:   true,
	}

	_, o, _, _ := bash.RunWithReturn()
	bash.PanicIfError()
	output := strings.Split(o, "\n")
	output = output[2:]
	ret := make(map[string]string)
	for _, item := range output {
		if item != "" {
			blk := strings.Fields(item)
			ret[blk[0]] = blk[1]
		}
	}
	return ret
}

func vmBlkStatusHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmStatusCheckCmd{}
	ctx.GetCommand(cmd)
	vm_status := make(map[string]map[string]string)
	for _, uuid := range cmd.Uuids {
		vm_status[uuid] = vmBlkStatus(uuid)
	}

	return VmBlkStatusRsp{Vm_status: vm_status}
}

func vmEchoHandler(ctx *server.CommandContext) interface{} {
	return nil
}

func vmDeviceQosHandler(ctx *server.CommandContext) interface{} {
	cmd := &VmDiskQosCmd{}
	ctx.GetCommand(cmd)
	bash := utils.Bash{
		Command: fmt.Sprintf("virsh blkdeviotune %s %s", cmd.Uuid, cmd.Dev),
		NoLog:   true,
	}

	_, o, _, _ := bash.RunWithReturn()
	bash.PanicIfError()

	return VmDiskQosRsp{VmDeviceIo: o}
}

func VmEntryPoint() {
	server.RegisterSyncCommandHandler(IS_VM_STOPPED_PATH, vmIsVmStoppedHandler)
	server.RegisterSyncCommandHandler(IS_VM_PAUSED_PATH, vmIsVmSuspendedHandler)
	server.RegisterSyncCommandHandler(IS_VM_DESTROYED_PATH, vmIsVmDestroyedHandler)
	server.RegisterSyncCommandHandler(IS_VM_RUNNING_PATH, vmIsVmRunningHandler)
	server.RegisterSyncCommandHandler(SSH_GUEST_VM_PATH, vmSshGuestVmHandler)
	server.RegisterSyncCommandHandler(SCP_GUEST_VM_PATH, vmScpGuestVmHandler)
	server.RegisterSyncCommandHandler(VM_STATUS, vmStatusHandler)
	server.RegisterSyncCommandHandler(LIST_ALL_VM, vmListAllVmHandler)
	server.RegisterSyncCommandHandler(VM_BLK_STATUS, vmBlkStatusHandler)
	server.RegisterSyncCommandHandler(VM_ECHO_PATH, vmEchoHandler)
	server.RegisterSyncCommandHandler(VM_DEVICE_QOS, vmDeviceQosHandler)
}
